runtime/drivers/file/model_executor_olap_self.go

```diff
@@ -328,22 +339,54 @@ func writeParquet(res *drivers.Result, path string) error {
 				recordBuilder.Field(i).(*array.BinaryBuilder).Append(v)
 			}
 		}
+		rows++
+		// We write the data to the underlying file to not buffer entire data in memory till it is written to parquet file.
+		// Since the check is not on memory but on number of rows this can still consume lot of memory
+		// so we set the limit to a much aggressive value than the default which is 64M.
+		if rows == maxParquetRowGroupLen {
+			rec := recordBuilder.NewRecord()
+			if err := parquetwriter.Write(rec); err != nil {
+				rec.Release()
+				return err
+			}
+			rec.Release()
+		}
 	}
 	if res.Err() != nil {
 		return res.Err()
 	}
-
-	f, err := os.Create(path)
-	if err != nil {
-		return err
+	if rows%maxParquetRowGroupLen == 0 {
+		return nil
 	}
-	defer f.Close()
+	rec := recordBuilder.NewRecord()
+	err = parquetwriter.Write(rec)
+	// release the record before returning the error
+	rec.Release()
+	return err
+}
 
-	parquetwriter, err := pqarrow.NewFileWriter(schema, f, nil, pqarrow.ArrowWriterProperties{})
-	if err != nil {
-		return err
-	}
-	defer parquetwriter.Close()
+// A limitedWriter writes to W but limits the amount of
+// data written to just N bytes.
+//
+// Copied from github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/util/ioutils/ioutils.go
+type limitedWriter struct {
+	W io.Writer // underlying writer
+	N int64     // max bytes remaining
+}
 
-	return parquetwriter.Write(recordBuilder.NewRecord())
+func (l *limitedWriter) Write(p []byte) (n int, err error) {
+	if l.N <= 0 {
+		return 0, io.ErrShortWrite
+	}
+	truncated := false
+	if int64(len(p)) > l.N {
+		p = p[0:l.N]
+		truncated = true
+	}
```

Comment:
Why does it do truncation? Shouldn't it just error without writing at all?

Reply:
I copied it from k8s source code and I could think of two reasons why they write the remaining bytes first and then fail. 
1. To confirm to the definition that the `limitedWriter` limits the data written to N bytes.
2. To give it a chance to underlying writer to return error if any till N bytes.

Reply:
Maybe those guarantees are important for how it is used in Kubernetes. Up to you, but since it is as a util writer in this specific package, I would remove it and save the extra writes as soon as we know the limit is breached.